### PR TITLE
jar lookup of manifest and version details

### PR DESCRIPTION
**Pull request by jodygarnett**
Used to verify geotools execution environment.

### CODE REVIEW

**Comment by jodygarnett** [comment]
@tbarsballe could you review please, adding this functionality to GeoTools next to other jar and environment maintenance methods.

**Comment by tbarsballe** [comment]
Looks good overall.
One thing I am wondering is how this would behave if running from an unexploded war. It looks like it should work fine, but that seems like something you might want to test manually.

File: modules/library/metadata/src/main/java/org/geotools/factory/GeoTools.java [hunk 1]
```diff
@@ -376,13 +381,103 @@ public static String getBuildTimestamp() {
...
+     * <li>build.timestamp=21-Mar-2016 21:30</li>
+     * </ul>
      */
     public static Properties getBuildProperties() {
         Properties props = new Properties();
         props.putAll(PROPS);
         return props;
     }
 
+    public static Version getVersion(Class<?> type) {
```
Comment (tbarsballe) [inline modules/library/metadata/src/main/java/org/geotools/factory/GeoTools.java]: Missing javadoc

File: modules/library/metadata/src/test/java/org/geotools/factory/GeoToolsTest.java [hunk 2]
```diff
@@ -20,8 +20,19 @@
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
```
Comment (tbarsballe) [inline modules/library/metadata/src/test/java/org/geotools/factory/GeoToolsTest.java]: ^ Headers say 2007-2008 ?

**Comment by jodygarnett** [comment]
I expect this will fail from an unexploded war - since  the protocol would either be custom (ie not jar) or zip. The protocol will fail in a stable way, returning null for "unknown" similar to how it behaves in an IDE.

**Comment by tbarsballe** [comment]
Yes, looks like this will not work in JBOSS - their paths look like `vfs:/var/jboss/workspace/BuildSvr_FNMOC/jboss/geoserver/deployments/geoserver.war/WEB-INF/lib/gt-xsd-wcs-13.2.jar/org/geotools/wcs/getCoverage.xsd`

(Custom protocol + no `!`)

See https://osgeo-org.atlassian.net/browse/GEOS-7139

**Comment by jodygarnett** [comment]
I could try search for ".war/" to catch the jboss case, another option is to "count backwords the correct number of directories based on the package name. I will have to factor out the method to make it testable, but yes let me cover that case and we can merge.

**Comment by jodygarnett** [comment]
@tbarsballe Please see recent commit which supports your jboss use-case - factored out non public utility methods to make the logic testable.

**Comment by tbarsballe** [comment]
Looks good, I am quite happy with this now.

**Pull request merged**